devolutions-session/src/dvc/window_monitor.rs

```diff
@@ -0,0 +1,635 @@
+    let send_result = unsafe {
+        SendMessageTimeoutW(
+            hwnd,
+            WM_GETTEXT,
+            WPARAM(MAX_TITLE_LENGTH),
+            LPARAM(title_buffer.as_mut_ptr() as isize),
+            SMTO_ABORTIFHUNG,
+            DEFAULT_SEND_MESSAGE_TIMEOUT_MS,
+            Some(&mut result),
+        )
+    };
+
+    if send_result.0 != 0 && result > 0 {
+        // Convert UTF-16 to String, removing null terminator.
+        let title = String::from_utf16_lossy(&title_buffer[..result]);
+        return Ok(title);
+    }
+
+    // No title available.
+    Ok(String::new())
+}
+
+/// Gets the process ID for a given window.
+fn get_window_process_id(hwnd: HWND) -> Result<u32> {
+    if hwnd.is_invalid() {
+        bail!("invalid window handle");
+    }
+
+    let mut process_id: u32 = 0;
+
+    // SAFETY: FFI call with no outstanding preconditions.
+    unsafe {
+        GetWindowThreadProcessId(hwnd, Some(&mut process_id));
+    }
+
+    if process_id == 0 {
+        bail!("failed to get process ID for window");
+    }
+
+    Ok(process_id)
+}
+
+/// Captures a snapshot of window information for a given window handle.
+fn capture_window_snapshot(hwnd: HWND) -> Result<WindowSnapshot> {
+    if hwnd.is_invalid() {
+        bail!("invalid window handle");
+    }
+
+    let title = get_window_title(hwnd).context("failed to get window title")?;
+    let process_id = get_window_process_id(hwnd).context("failed to get process ID")?;
+
+    // Open process to query information.
+    // This may fail for protected/system processes, which is expected.
+    let process = Process::get_by_pid(process_id, PROCESS_QUERY_INFORMATION)
+        .context("failed to open process (may be protected/system process)")?;
+
+    let exe_path = process.exe_path().context("failed to get executable path")?;
+
+    Ok(WindowSnapshot {
+        process_id,
+        title,
+        exe_path,
+    })
+}
+
+/// Captures information about the currently active window.
+fn capture_foreground_window() -> Result<WindowSnapshot> {
+    // SAFETY: FFI call with no outstanding preconditions.
+    let foreground_window = unsafe { GetForegroundWindow() };
+
+    if foreground_window.is_invalid() {
+        bail!("no foreground window");
+    }
+
+    capture_window_snapshot(foreground_window)
+}
+
+/// Gets the current timestamp as seconds since Unix epoch.
+fn get_current_timestamp() -> u64 {
+    SystemTime::now()
+        .duration_since(std::time::UNIX_EPOCH)
+        .map(|d| d.as_secs())
+        .unwrap_or(0)
+}
+
+/// Thread-local context for the event hook callback.
+///
+/// Windows event hook callbacks must be plain C functions, so we use thread-local
+/// storage to communicate with the async runtime. This is safe because WINEVENT_OUTOFCONTEXT
+/// ensures the callback runs on the same thread that installed the hook.
+struct HookContext {
+    sender: mpsc::UnboundedSender<WindowSnapshot>,
```

Comment:
nitpick: Unbounded channels are typically a yellow/red-ish flag. If you really need one, document why it’s really better than using a bounded channel with proper backpressure.